sklearn/utils/extmath.py

```diff
@@ -255,37 +257,71 @@ def randomized_range_finder(
     analysis
     A. Szlam et al. 2014
     """
+    xp, is_array_api_compliant = get_namespace(A)
     random_state = check_random_state(random_state)
 
     # Generating normal random vectors with shape: (A.shape[1], size)
-    Q = random_state.normal(size=(A.shape[1], size))
-    if hasattr(A, "dtype") and A.dtype.kind == "f":
-        # Ensure f32 is preserved as f32
-        Q = Q.astype(A.dtype, copy=False)
+    # XXX: generate random number directly from xp if it's possible
+    # one day.
+    Q = xp.asarray(random_state.normal(size=(A.shape[1], size)))
+    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
+        # Use float32 computation and components if A has a float32 dtype
+        Q = xp.astype(Q, A.dtype, copy=False)
+
+    # Move Q to device if needed only after converting to float32 if needed to
+    # avoid allocating unnecessary memory on the device.
+    if hasattr(A, "device"):
+        if not hasattr(Q, "to_device"):
+            # array_api_compat for PyTorch does not seem to make it possible
+            # to use the intended to_device API:
+            # https://data-apis.org/array-api/2022.12/design_topics/device_support.html
```

Comment:
The `array_api_compat` [README](https://github.com/data-apis/array-api-compat/tree/main#helper-functions) says that they have a `to_device` helper function to deal with array from libs that do not have `x.to_device` (torch, cupy, numpy). Should we use that instead of `xp.asarray(.., device=A.device)`?

Reply:
There's also a `device` function to get rid of the other hasattr (I'm curious why you use `hasattr(A, "dtype")`. Is that something we also need a helper for?)

Reply:
FWIW, this code got me thinking a little bit about the difference between `to_device` and `asarray`. One difference right now is that `asarray` lets you be explicit about copying or not when the array is already on the same device, but I think this is just an underspecification in the standard right now. https://github.com/data-apis/array-api/issues/645

I didn't open an issue for it, but perhaps it would be useful if `to_device` had a dtype argument, similar to `torch.to`, to let you do this whole operation in one step (although I'm honestly surprised that the numpy random generators don't let you natively generate float32).